bpo-31589: Add config for LaTeX handling of stray Unicode chars in PDF

jfbu opened the pull request:
Does not modify config for xelatex, lualatex or platex (Japanese).


<!-- issue-number: bpo-31589 -->
https://bugs.python.org/issue31589
<!-- /issue-number -->

jfbu committed:
bpo-31589: Add config for LaTeX handling of stray Unicode chars in PDF

Does not modify config for xelatex, lualatex or platex (Japanese).

the-knights-who-say-ni commented:
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately our records indicate you have not signed the CLA. For legal reasons we need you to sign this before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

JulienPalard commented:
This would greatly simplify https://github.com/python/docsbuild-scripts/pull/34.

I also find it better to have all necessary configuration in one place instead of scatered between cpython and docsbuild-scripts, depending on docsbuild-scripts to build documentation translations is not a good thing.

JulienPalard commented:
I ran a full test build with the build_docs.py script, and it went very well, just had fails on 2.7 due to translation errors we fixed for 3.6 but not backported (like U+200B characters in translations).

Note: this strict configuration permits to find those bugs (bugs like U+200B in translations) which is nice. In the other hand, adding a new unicode character needs to "whitelist" it manually in conf.py like https://github.com/python/cpython/pull/4069/files#diff-a96b84821bf04e0f0bf3c216ee1cfb92R110 which does not look to happen often as there's only 4 currently listed.

jfbu commented on Doc/conf.py:
```diff
@@ -92,19 +92,26 @@
+\ifx\kanjiskip\undefined          % not for platex
+\usepackage{newunicodechar}
+\newunicodechar{ſ}{{\fontencoding{TS1}\fontfamily{lmr}\selectfont s}}
+\newunicodechar{K}{\ensuremath{\mathrm K}}
```
@JulienPalard in this line when one copy pastes from inside Firefox from the github web view one gets a standard ascii K in as first argument to `\newunicodechar`. But in my commit it really is U+212A (KELVIN SIGN).

JulienPalard replied:
Thanks for notifying. I used curl | git apply if I remember correcly and everything is fine. Errors I still have come from translations strings on 2.7 so we should just fix them in our translation repositories.

jfbu replied:
I got trapped myself because I copied pasted from the web view directly to `Doc/conf.py` rather than git applying my patch to another branch.

JulienPalard commented:
Ran a test using docsbuild-scripts.

Had to manually fix a typo in the 2.7 branch of the japanese translation (https://github.com/python/python-docs-ja/issues/3), once fixed all builds are successfull, so LGTM.

jfbu commented:
I had actually signed the CLA before (i.e. minutes before) submitting the PR and I don't know how to make the CLA not signed tag go away.

JulienPalard commented:
@jfbu it take a manual action, but I don't have the rights to trigger it I think.

JulienPalard commented:
@jfbu Hi, is there a simple procedure to find how to write a `newunicodechar{…}`?

I tried using the config resulting of this PR for a documentation of mine and got some errors about other characters like №, ×, maybe €, ... And I were unable to forge a correponding `newunicodechar{№}`.

jfbu commented:
@JulienPalard unfortunately I couldn't describe a simple procedure working all the time. What I would do is use `utf8x` option to inputenc to see what happens:

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8x]{inputenc}
\begin{document}
№, ×, €
\end{document}
```

Then I try pdflatex this document. There are errors `\textnumero undefined`, and `\texteuro undefined`. These errors are more explicit than those which would come from `utf8`, which would say `Unicode char № (U+2116) not set-up for LaTeX` only. I am aware there is some package `textcomp` which provides additional symbols, so I try again with 

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8x]{inputenc}
\usepackage{textcomp}
\begin{document}
№, ×, €
\end{document}
```

and it all works. Then I try my luck again with `utf8` 

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp}
\begin{document}
№, ×, €
\end{document}
```

and it works... Does adding `\usepackage{textcomp}` to your preamble solve your issues ?

**edit** make sure to read bottom of this before trying...

You can find the additional Unicode code-points it defines in file `ts1enc.dfu` (`kpsewhich ts1enc.dfu` returns `/usr/local/texlive/2017/texmf-dist/tex/latex/base/ts1enc.dfu` on my system). Perhaps Sphinx should do `\usepackage{textcomp}` per default.

This does not quite answer your question; as `utf8x` (which works with package `ucs`) has extensive support files, I sometimes have to dig into them to find out which font encoding and which font slot I should use in `\newunicodechar`. For example imagine I am looking for ℂ which is U+2102. 

- I convert 0x2102 to decimal 8450

- I move to `ucs` repertory in my TeX distribution and grep 8450 there

```
$ kpsewhich ucs.sty
/usr/local/texlive/2017/texmf-dist/tex/latex/ucs/ucs.sty

$ pushd /usr/local/texlive/2017/texmf-dist/tex/latex/ucs
/usr/local/texlive/2017/texmf-dist/tex/latex/ucs ~/_texlatex/1711

$ grep -r 8450
data/uni-111.def:\uc@dclc{28450}{cjkbg5}{\u@cjk@Bgv1693}%
data/uni-111.def:\uc@dclc{28450}{cjkjis}{\jischar{3441}}%
data/uni-150.def:\uc@dclc{38450}{cjkbg5}{\u@cjk@Bgv05A7}%
data/uni-150.def:\uc@dclc{38450}{cjkgb}{\u@cjk@GB0933}%
data/uni-150.def:\uc@dclc{38450}{cjkjis}{\jischar{4B49}}%
data/uni-250.def:\uc@dclc{64071}{autogenerated}{\unichar{28450}}%
data/uni-250.def:\uc@dclc{64154}{autogenerated}{\unichar{28450}}%
data/uni-33.def:\uc@dclc{8450}{default}{\ensuremath{\mathbb C}}%

```

There are false-positive but I find the definition `\ensuremath{\mathbb C}`. Thus I can do

```latex
\newunicodechar{ℂ}{\ensuremath{\mathbb C}}
```

(the ams packages loaded by Sphinx provide `\mathbb` blackboard alphabet -- this is just an example).

When I see that the utf8x defintion would use some `\text...` macro, I try my luck with `textcomp` package. Or I should have tried that first...

I understand the whole thing is a bit scary. And then we have additional problem:

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp}
\usepackage{times}% default for Sphinx
\begin{document}
№, ×, €
\end{document}
```

gives error:

```
./temp77.tex:8: Package textcomp Error: Symbol \textnumero not provided by
(textcomp)                font family ptm in TS1 encoding.
(textcomp)                Default family used instead.
```

which is ridiculous because this should be only a warning, not an error. It said that it had to use computer modern, not Times font. Which means to avoid that error we must do

```latex
\newunicodechar{№}{{\fontfamily{cmr}\selectfont\textnumero}}
```

Final mwe

```latex
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp}
\usepackage{times}
\usepackage{newunicodechar}
\newunicodechar{№}{{\fontfamily{cmr}\selectfont\textnumero}}
\begin{document}

№, ×, €
%\showoutput
\end{document}
```

...whow `:-(` ... well we did it...

jfbu commented:
Actually my previous comment did not describe what if the `utf8x` test had compiled with no error to start with... I would have switched to the grep method to examine the support files as described in my comment for the case of blackboard C. (which we could have guessed, but that was only example).

JulienPalard commented:
Thanks for this extensive answer, I'm not sure this is the right way but I'm learning a lot about latex.

I'm trying to follow it step by step, first build without changing nothing I'm getting:

```
! Package textcomp Error: Symbol \textnumero not provided by
(textcomp)                font family ptm in TS1 encoding.
(textcomp)                Default family used instead.

See the textcomp package documentation for explanation.
```

So I'm trying with utf8x, I'm getting the same error, I'm tring with textcomp, without utf8x I'm getting the same error, with utf8x I'm getting the same error.

I'm following your answer and I'm grepping for № in ucs.sty which gives:

    /usr/share/texlive/texmf-dist/tex/latex/ucs/data/uni-33.def:\uc@dclc{8470}{default}{\textnumero}%

So I'm trying with:

    \newunicodechar{№}{\textnumero}

Which expectedly yields:

    ! Package textcomp Error: Symbol \textnumero not provided by
    (textcomp)                font family ptm in TS1 encoding.
    (textcomp)                Default family used instead.

So I'm trying with your version:

    \newunicodechar{№}{{\fontfamily{cmr}\selectfont\textnumero}}

and the error is now gone.

Next error is:

```
! Package inputenc Error: Unicode char ⅓ (U+2153)
(inputenc)                not set up for use with LaTeX.

See the inputenc package documentation for explanation.
Type  H <return>  for immediate help.
 ...                                              
                                                  
l.1056 50\% dans l’autre. Voir ⅓
```

`U+2153` is `VULGAR FRACTION ONE THIRD`, Decimal: `&#8531`, so I'm grepping:

wich gives:

    /usr/share/texlive/texmf-dist/tex/latex/ucs/data/uni-33.def:\uc@dclc{8531}{autogenerated}{\unichar{49}\unichar{8260}\unichar{51}}%

So I'm trying:

    \newunicodechar{⅓}{\unichar{49}\unichar{8260}\unichar{51}}

But it gives:

    ! Undefined control sequence.
    \u8:⅓ ->\unichar 
                       {49}\unichar {8260}\unichar {51}

I think I have failed building the newunicodechar given the result of my grep, probably a bad syntax, I'm not fluent enough in latex to spot it :(

I don't think documentation people are willing to learn this, even if it's super interesting and nice to learn, it's a whole other subject.

The only way I see which may make this almost OK would be to tell documentation people not to care about PDF builds (I think they only build HTML localy so they already don't really care about PDF they don't even have to know that latex is used to build PDFs), and to clearly document how to fix build errors, this way we have two distincts teams with two distincts sets of knowledge, one caring and focusing about redacting nice documentation, and the other caring about PDF builds. Looks a bit  over engineered to me.

jfbu commented:
The `\unichar` is a `ucs` command. Sphinx is compatible with doing `\usepackage[utf8x]{inputenc}` (which loads `ucs`) rather than `\usepackage[utf8]{inputenc}` but I dimly remember this raised other issues in your context.

`\unichar` simply means the character with that decimal code-point.

So `\newunicodechar{⅓}{\unichar{49}\unichar{8260}\unichar{51}}` simply means `\newunicodechar{⅓}{1⁄3}` where I used the FRACTION SLASH U+2044. By luck it does not need extra definition with the already obtained set-up.

```
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp}
\usepackage{times}
\usepackage{newunicodechar}
\newunicodechar{№}{{\fontfamily{cmr}\selectfont\textnumero}}
\newunicodechar{⅓}{1⁄3}

\begin{document}
№, ×, €, ⅓
\end{document}
```

The result isn't very good for `⅓`, I am sure there are LaTeX packages for that (I think there actually is one developed by the LaTeX3 team, yes it is called `xfrac`.)

About your other comments, well, yes supporting Unicode with pdflatex is a rough path. Sphinx projects definitely do provide the possibility that user configures it to go via XeLaTeX or LuaLaTeX and suitable fonts. You don't have to go via pdflatex. But if you go via XeLaTeX or LuaLaTeX you need to choose fonts with wide enough Unicode coverage.

jfbu commented:
Here is with `xfrac` package.

```
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage{textcomp}
\usepackage{times}
\usepackage{newunicodechar}
\newunicodechar{№}{{\fontfamily{cmr}\selectfont\textnumero}}
\usepackage{xfrac}
\newunicodechar{⅓}{\sfrac{1}{3}}

\begin{document}
№, ×, €, ⅓
\end{document}
```

output

![capture d ecran 2017-11-29 a 15 07 03](https://user-images.githubusercontent.com/2589111/33379015-0da88882-d517-11e7-8827-2e645d2acb50.png)

For more info https://tex.stackexchange.com/questions/3885/how-to-get-a-little-frac

jfbu commented:
sorry I meant https://tex.stackexchange.com/questions/3372/how-do-i-typeset-arbitrary-fractions-like-the-standard-symbol-for-5-%C2%BD

jfbu commented:
About my remark

> I dimly remember this raised other issues in your context.

the № is a case in point. It is supported by `utf8x` option, but one has to add the `\usepackage{textcomp}`. But then it does not work with `\usepackage{times}` because the font does not provide the glyph. So we have to do now

    \DeclareUnicodeCharacter{"2116}{{\fontfamily{cmr}\selectfont\textnumero}}

Notice that it is very regrettable that the `\DeclareUnicodeCharacter` with `utf8x` behaves differently from the `\DeclareUnicodeCharacter` with `utf8` (which is the LaTeX team supported one). The former accepts either a decimal number (here `8470`)  or an hexadecimal number in TeX notation, so with `"` prefix, (or an octal number...), but the latter, the LaTeX team one accepts **only** hexadecimal code-point `2116` with no prefix `:-(`. (to be clear, I consider the `utf8x` one much better, and it is a pain they both have exactly the same macro name but behave differently ...)

The package `newunicodechar` refuses absolutely to work with `utf8x`. So we have to use the code-point of the unicode character and proceed as above.

It is not either an out-of-the-box solution, no question about that. It is possible that the now wide-spread use of the Unicode engines has stopped a bit the documentation for general LaTeX user on how to survive really with pdflatex + Unicode. Notice that `utf8` option had a very limited coverage of Unicode initially, but the LaTeX team has extended it these last two or three years, an  indirect indication that there are too many often encountered problems for general user. But they proceed by very conservative steps and as I said whenever you have a Unicode question nowadays, most everyone will tell you to use xelatex or lualatex. Which however does not solve all (one may need to switch language so that the correct font is used for the given glyph; even there it is NOT all automatic for LaTeX user.)

jfbu commented:
I concur with your conclusion you need to set-up a PDF task force. And if you do, please report your findings to the Sphinx maintainers, for example if you switch to xelatex, which fonts do you use to your satisfaction allowing simultaneous builds into French, Japanese, or Hebrew.

edit: or rather the occasional use of an isolated Unicode codepoint for whatever reason. I am interested into any robust solution.

jfbu commented:
@JulienPalard  I realize only now that Sphinx already loads package `textcomp` (I should have checked earlier). Hence, all the discussion prior to ⅓ would not have arisen for a Sphinx project were it not for the fact that `textcomp` raises an error for № because it can't be find it in times font per default.

**But** there is a package option (I had to dig into latex source code for finding it) `warn` which converts this error into a simple warning!

```
\documentclass{article}
\usepackage[T1]{fontenc}
\usepackage[utf8]{inputenc}
\usepackage[warn]{textcomp}
\usepackage{times}
\begin{document}
№, ×, €, ½ % but ⅓ needs extra as it is not available in TeX font encodings I
           % know of.

% ⅓
\end{document}
```

works with no further ado. (regarding the € because textcomp knows that times font supports the `\texteuro` command). I will make a PR at Sphinx to pass option `warn` to textcomp.

jfbu committed:
Add NEWS file entry via blurb add

	new file:   Misc/NEWS.d/next/Documentation/2017-12-06-20-01-22.bpo-31589.ystCoY.rst

jfbu committed:
Better not set latex's 'fontenc' key

Indeed, this way the conf.py can have an added latex_engine = 'xelatex'
if desired with no other change; the pdflatex config added by these
commits will remain invisible to xelatex case.

The \PassOptionsToPackage{warn}{textcomp} will be un-needed with Sphinx
1.6.6 or later.

	modified:   Doc/conf.py
